Barycentric_coordinates_3/include/CGAL/Barycentric_coordinates_3/Wachspress_coordinates_3.h

```diff
@@ -0,0 +1,398 @@
+    const GeomTraits m_traits;
+
+    const Dot_3 m_dot_3;
+    const Det_3 m_det_3;
+    const Cross_3 m_cross_3;
+    const Construct_vec_3 m_construct_vector_3;
+
+    std::vector<FT> m_weights;
+
+    template<typename OutputIterator>
+    OutputIterator compute(
+      const Point_3& query, OutputIterator coordinates) {
+
+      switch(m_computation_policy){
+
+        case Computation_policy_3::FAST:{
+          return compute_coords(query, coordinates);
+        }
+
+        case Computation_policy_3::FAST_WITH_EDGE_CASES:{
+          // Calculate query position relative to the polyhedron
+          const auto edge_case = internal::locate_wrt_polyhedron(
+            m_vertex_to_point_map, m_polygon_mesh, query, coordinates, m_traits, true);
+
+          if(edge_case == internal::Edge_case::BOUNDARY) {
+            return coordinates;
+          }
+          if(edge_case == internal::Edge_case::EXTERIOR_BOUNDARY){
+            std::cerr << std::endl <<
+            "WARNING: query does not belong to the polygon!" << std::endl;
+            return coordinates;
+          }
+          if(edge_case == internal::Edge_case::EXTERIOR) {
+            std::cerr << std::endl <<
+            "WARNING: query does not belong to the polygon!" << std::endl;
+          }
+
+          return compute_coords(query, coordinates);
+        }
+
+        default:{
+          internal::get_default(vertices(m_polygon_mesh).size(), coordinates);
+          return coordinates;
+        }
+      }
+      return coordinates;
+    }
+
+    template<typename OutputIterator>
+    OutputIterator compute_coords(
+      const Point_3& query, OutputIterator coordinates){
+
+      // Compute weights.
+      const FT sum = compute_weights(query);
+      CGAL_assertion(sum != FT(0));
+
+      // The coordinates must be saved in the same order as vertices in the vertex range.
+      const auto vd = vertices(m_polygon_mesh);
+      CGAL_assertion(m_weights.size() == vd.size());
+
+      for (std::size_t vi = 0; vi < vd.size(); vi++) {
+
+        CGAL_assertion(vi < m_weights.size());
+        const FT coordinate = m_weights[vi]/sum;
+        *(coordinates++) = coordinate;
+      }
+
+      return coordinates;
+    }
+
+    FT compute_weights(const Point_3& query) {
+      // Sum of weights to normalize them later.
+      FT sum = FT(0);
+
+      // Vertex index.
+      std::size_t vi = 0;
+      for (auto vertex : vertices(m_polygon_mesh)) {
+
+        // Call function to calculate wp coordinates
+        const FT weight = compute_wp_vertex_query(vertex, query);
+
+        CGAL_assertion(vi < m_weights.size());
+        m_weights[vi] = weight;
+        sum += weight;
+        ++vi; // update vi
+      }
+
+      CGAL_assertion(sum != FT(0));
+      return sum;
+    }
+
+    // Compute wp coordinates for a given vertex v and a query q
+    template<typename Vertex>
+    FT compute_wp_vertex_query(const Vertex& vertex, const Point_3& query){
```

Comment:
@sloriot The issue is probably coming from this method. There is suspicious negation of the first face normal, which is not coherent with other normals used here. Implementations of other MV and DH coordinates do not have this. My guess would be that this method does not return the same weights for the same mesh but differently oriented faces, which should not be the case. What is important here is to have consistent normals as indicated in the [original paper](https://www.cs.wustl.edu/~taoju/research/generalbary_new.pdf). Maybe just try to remove the negation in line 269 and see what happens? 

There are multiple formulas to compute WP coordinates. I do not remember which one Antonio used here. Maybe it is worth using all formulas from the same paper to be consistent. Finally, another test that is easy to do is to change WP to DH coordinates in this test and see if it works. If I remember correctly for such a simple case (center point in a cube) WP and DH should produce exactly the same result.